### DISCUSSION TITLE
Limiting number of nodes for each vehicle

### DISCUSSION THREAD

**Discussion by alti3** [google/or-tools#3181] [Routing (and legacy CP) questions] [answered]
I'm solving a VRP with capacity constraints and pickup/delivery constraints.
I don't want any vehicle to be assigned a route longer than a certain value, so I added a **Distance Constraint** as shown in the docs

```csharp
// Add Distance constraint.
routing.AddDimension(
  transitCallbackIndex,
  0,
  3000,
  true,
  // start cumul to zero
  "Distance");
RoutingDimension distanceDimension = routing.GetMutableDimension("Distance");
distanceDimension.SetGlobalSpanCostCoefficient(100);
```

While the above code works great, there are some edge cases where many pickups/deliveries are geographically close, leading to the undesired scenario where a single vehicle is assigned too many nodes, while still satisfying the vehicle capacity and distance constraints. The solver achieves this by assigning the vehicle several pickups (say 3 pickups for example) followed by a series of deliveries to offload the demand of the first 3 pickups, followed by another series of pickups and deliveries.

In one particular case, a vehicle was assigned 11 nodes along its route (starting node + 5 pickups + 5 deliveries).

Reducing the maximum value of the distance constraint doesn't solve this problem simply because the pickups/deliveries are in close proximity.

I understand the best solution is by creating a new constraint for the number of nodes assigned to a single vehicle.
How can this be done?

**Comment by vuongdanghuy** [comment, marked as answer]
I think you can create a new capacity dimension for the number of nodes assigned to each vehicle and each node will has a demand equal to 1.
For example:
```python
routing.AddDimensionWithVehicleCapacity(
    callback_index, 
    0,
    [5, 6, 7, 8] // maximum number of nodes
    true, // start cumul to zero
    "Nodes");
```
That will limit the maximum number of nodes for each vehicle to 5, 6, 7 and 8 respectively.

Hope this can help.

**Reply by alti3** [reply]
> I think you can create a new capacity dimension for the number of nodes assigned to each vehicle and each node will has a demand equal to 1. For example:
> 
> ```python
> routing.AddDimensionWithVehicleCapacity(
>     callback_index, 
>     0,
>     [5, 6, 7, 8] // maximum number of nodes
>     true, // start cumul to zero
>     "Nodes");
> ```
> 
> That will limit the maximum number of nodes for each vehicle to 5, 6, 7 and 8 respectively.
> 
> Hope this can help.

Thanks so much

Another question:

Am I right in assuming that the the max number of nodes should always be an odd number?
My thought was that pickup/delivery nodes are pairs and their count will always be even, and when you add the vehicle starting node, the total will always be odd.

If I set the maximum number of nodes to be even, there's a chance that the solver might assign the pickup node to a particular vehicle while dropping the corresponding delivery node (I have allowed drops and set penalties for each node in the matrix).

Thanks in advance

**Reply by vuongdanghuy** [reply]
As far as I know, if the delivery node is dropped due to capacity constraint, the pick up node will be dropped as well. So it does not matter that the maximum number of nodes is even or odd, you just need to pick an appropriate number that satisfied your problem.

**Reply by alti3** [reply]
> As far as I know, if the delivery node is dropped due to capacity constraint, the pick up node will be dropped as well. So it does not matter that the maximum number of nodes is even or odd, you just need to pick an appropriate number that satisfied your problem.

That seems to be the case indeed.
I just tried both odd and even max-node limits and it doesn't seem to make a difference.
The solver is handling this internally when a delivery node is dropped, its corresponding pickup node is also dropped.

Works like a charm
Thanks a lot, much appreciated
